File: static/solutions/java/src/main/java/com/epi/BinarySearchFirstK.java
package com.epi;

/*
    @slug
    first-occurence-of-k

    @title
    Search a sorted array for first occurrence of a key

    @problem
    Write a method that takes a sorted array and a key and returns the index of
   the
    first occurrence of that key in the array. For example, when applied to the
   array in
    the figure, your algorithm should return 3 if the given key is 108; if it is
   285, your
    algorithm should return 6.
    <p>

    <img src="/bin-search-first.png" width="500px"></img>
    <p>

    @hint
    What happens when every entry equals k? Don't stop when you first see k.

 */

import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class BinarySearchFirstK {
  // @include
  // @judge-include-display
  public static int searchFirstOfK(List<Integer> A, int k) {
    // @judge-exclude-display
    int left = 0, right = A.size() - 1, result = -1;
    // A.subList(left, right + 1) is the candidate set.
    while (left <= right) {
      int mid = left + ((right - left) / 2);
      if (A.get(mid) > k) {
        right = mid - 1;
      } else if (A.get(mid) == k) {
        result = mid;
        // Nothing to the right of mid can be the first occurrence of k.
        right = mid - 1;
      } else { // A.get(mid) < k
        left = mid + 1;
      }
    }
    return result;
    // @judge-include-display
  }
  // @judge-exclude-display
  // @exclude

  private static int checkAns(List<Integer> A, int k) {
    for (int i = 0; i < A.size() && A.get(i) <= k; ++i) {
      if (A.get(i) == k) {
        return i;
      }
    }
    return -1;
  }

  private static void check(int expected, int got, List<Integer> A, int key) {
    if (expected != got) {
      System.err.println("Failed looking for " + key + " in array " + A);
      System.err.println("Got " + got + ", expected " + expected);
      System.exit(-1);
    }
  }

  private static void simpleTest() {
    List<Integer> A = Arrays.asList(0, 1, 2, 3, 4, 5, 6, 7);
    check(0, searchFirstOfK(A, 0), A, 0);
    check(1, searchFirstOfK(A, 1), A, 1);
    check(4, searchFirstOfK(A, 4), A, 4);
    check(6, searchFirstOfK(A, 6), A, 6);
    check(7, searchFirstOfK(A, 7), A, 7);
    check(-1, searchFirstOfK(A, 8), A, 8);
    check(-1, searchFirstOfK(A, Integer.MIN_VALUE), A, Integer.MIN_VALUE);
    A = Arrays.asList(1, 1, 2, 3, 4, 5, 6, 7);
    check(0, searchFirstOfK(A, 1), A, 1);
    A = Arrays.asList(1, 1, 2, 3, 4, 4, 4, 7);
    check(4, searchFirstOfK(A, 4), A, 4);
    A = Arrays.asList(1, 1, 1, 1, 1, 2);
    check(-1, searchFirstOfK(A, 0), A, 0);
    check(0, searchFirstOfK(A, 1), A, 0);
    check(5, searchFirstOfK(A, 2), A, 5);
    A = Arrays.asList(1, 1, 1, 1, 2, 2);
    check(4, searchFirstOfK(A, 2), A, 2);
  }

  public static void main(String[] args) {
    simpleTest();
    Random r = new Random();
    for (int times = 0; times < 1000; ++times) {
      int n;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
      } else {
        n = r.nextInt(10000) + 1;
      }
      List<Integer> A = new ArrayList<>();
      int k = r.nextInt(n);
      for (int i = 0; i < n; ++i) {
        A.add(r.nextInt(n));
      }
      Collections.sort(A);
      int ans = searchFirstOfK(A, k);
      System.out.println("k = " + k + " locates at " + ans);
      if (ans != -1) {
        System.out.println("A[k] = " + A.get(ans));
      }
      assert(checkAns(A, k) == ans);
    }
  }
}

File: static/python/binary_search_first_k.py
import sys
import random
import bisect


# @include
def search_first_of_k(A, k):
    left, right, result = 0, len(A) - 1, -1
    # A[left:right + 1] is the candidate set.
    while left <= right:
        mid = (left + right) // 2
        if A[mid] > k:
            right = mid - 1
        elif A[mid] == k:
            result = mid
            right = mid - 1  # Nothing to the right of mid can be solution.
        else:  # A[mid] < k.
            left = mid + 1
    return result
# @exclude


def simple_test():
    A = [0, 1, 2, 3, 4, 5, 6, 7]
    assert 0 == search_first_of_k(A, 0)
    assert 1 == search_first_of_k(A, 1)
    assert 4 == search_first_of_k(A, 4)
    assert 6 == search_first_of_k(A, 6)
    assert 7 == search_first_of_k(A, 7)
    assert -1 == search_first_of_k(A, 8)
    assert -1 == search_first_of_k(A, -2**64)
    A[0] = 1
    assert 0 == search_first_of_k(A, 1)
    A[5] = 4
    A[6] = 4
    assert 4 == search_first_of_k(A, 4)
    A = [1, 1, 1, 1, 1, 2]
    assert -1 == search_first_of_k(A, 0)
    assert 0 == search_first_of_k(A, 1)
    assert 5 == search_first_of_k(A, 2)
    A[4] = 2
    assert 4 == search_first_of_k(A, 2)


# Pythonic solution
def search_first_of_k_pythonic(A, k):
    i = bisect.bisect_left(A, k)
    return i if i < len(A) and A[i] == k else -1


def main():
    simple_test()
    for _ in range(1000):
        n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1,
                                                                       100000)
        k = random.randrange(n)
        A = sorted(random.randrange(n) for _ in range(n))
        ans = search_first_of_k(A, k)
        print('k =', k, 'locates at', ans)
        if ans != -1:
            print('A[k] =', A[ans])
        try:
            it = A.index(k)
        except ValueError:
            it = -1
        assert it == ans
        assert it == search_first_of_k_pythonic(A, k)


if __name__ == '__main__':
    main()
